Add workflow template for Facebook posts

jluis9796 opened the pull request:
- close #317 

This PR adds a workflow to post the blog posts in Facebook page.

Currently I added a template based on the Facebook API documentation and using a code suggested by GitHub Copilot.
As other workflow, I am using `actions/github-script`. However, I am not sure if this is the best approach since the scripts seems to be out of the scope of the code editor.

jluis9796 committed:
add workflow template for facebook posts

kenzauros reviewed (commented):
The basic idea is okay.

Basically split the phases (see `basic flow` in #317) of the implementation into separate steps.

1. rss.xml (https://mseeeen.msen.jp/rss.xml) を取得
2. RSS のハッシュを前回と比較して、同一なら、終了
3. RSS 内から最新の記事 (pubDate) を取得
4. Facebook に写真を投稿

Maybe you can use the cache system of Github Actions for the second step.

- [Caching dependencies to speed up workflows - GitHub Docs](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows)

And use latest features of JavaScript as much as possible.
(Like `fetch` API)

Additionally `workflow-dispatch` section should be placed into the file to run it manually.

jluis9796 committed:
add steps to facebook workflow

jluis9796 committed:
add branch to list

jluis9796 committed:
fix to fetch data

jluis9796 committed:
wrap string with quotes

jluis9796 committed:
modify fetch await

jluis9796 committed:
fix fetch response

jluis9796 committed:
fix function name

jluis9796 committed:
fix object key, set environment value

jluis9796 committed:
fix curl exec step

jluis9796 committed:
set step output encoding

jluis9796 committed:
fix condition to get date

jluis9796 committed:
modify access token setting

jluis9796 committed:
add step to get token

jluis9796 committed:
fix token variable

jluis9796 committed:
fix token API url

jluis9796 committed:
fix to read token from json

jluis9796 commented:
I wrote the basic steps and the workflow seems to run without problem (although some details may need improvement), but in the last step the token authentication is failing for some unknown reason with the error below.

www-authenticate: OAuth "Facebook Platform" "invalid_request" "An active access token must be used to query information about the current user."

![スクリーンショット 2024-05-09 154145](https://github.com/mseninc/blog/assets/10829398/de93daf4-3e6b-480a-ab33-09a92d3477cf)


I have tried to generate the access token using the FACEBOOK_APP_ID and the FACEBOOK_APP_SECRET based on this documentation
https://developers.facebook.com/docs/facebook-login/guides/access-tokens?locale=ja_JP


I thought the issue that I was using the method 
`curl -i -X GET "https://graph.facebook.com/{api-endpoint}&access_token={your-app_id}|{your-app_secret}" `
but I changed the steps to call the API to generate the token 
`curl -X GET "https://graph.facebook.com/oauth/access_token
  ?client_id={your-app-id}
  &client_secret={your-app-secret}
  &grant_type=client_credentials"
`
but I get the same error.

kenzauros commented:
I guess the error message says `This access token has no right to get information of the current user`.

Why don't you try to send a request to the url with Page ID `msenjp` instead of `me`?

kenzauros commented on .github/workflows/release-facebook.yml:
```diff
@@ -0,0 +1,82 @@
+        id: token
+        run: |
+          json=$(curl -X GET "https://graph.facebook.com/oauth/access_token?client_id=${{ vars.FACEBOOK_APP_ID }}&client_secret=${{ secrets.FACEBOOK_APP_SECRET }}&grant_type=client_credentials")
+          token=$(echo $json | grep -o '"access_token":"[^"]*' | grep -o '[^"]*$')
```
Is `jq` command not available?

kenzauros reviewed (commented):
By the way, I found a VS Code extension for GitHub Actions.

I haven't tried yet though it might be helpful.

- [GitHub Actions - Visual Studio Marketplace](https://marketplace.visualstudio.com/items?itemName=GitHub.vscode-github-actions)

jluis9796 commented:
Thanks for the suggestions. I'll try them.

jluis9796 committed:
fix page API url

jluis9796 committed:
set api version

jluis9796 committed:
add user input param

jluis9796 committed:
modify token setting

jluis9796 committed:
set token in the header

jluis9796 committed:
change url page

jluis9796 committed:
get user id

jluis9796 committed:
fix API params

jluis9796 committed:
debug token

jluis9796 committed:
modify token setting

jluis9796 committed:
fix header setting

jluis9796 committed:
change API page

jluis9796 committed:
fix header token

jluis9796 committed:
set API version

jluis9796 committed:
add dialog API

jluis9796 commented:
I am still failing to achieve the authentication.
I have read the documentation and blogs, and it happens "access token" actually is just an abbreviation. There are several types of access token. And the token I am generating may not to be the adequate one for creating posts.
https://developers.facebook.com/docs/facebook-login/guides/access-tokens/?locale=en_US

- App Access Token
- Client Token
- Page Access Token 
- System User Access Token
- User Access Token

The one I have been generating is the App Access Token. However, this token seems to be used for other purposes but not for posting. Instead, to create posts it seems I need a User Access Token or a Page Access Token.

Maybe I shoud ask you for a user information. In this case, I remember I read that the token generation requires javascript SDK.

In the case of Page Access Token, this blog showed how to generate it from the page.
https://yakuway.com/2020/08/08/1608/

I have found some blogs explaining similar setup, though I am not clear about their setup.

https://qiita.com/xiushu53/items/f2d699704d7325d02169
https://qiita.com/xiushu53/items/f2d699704d7325d02169
https://kantaro-cgi.com/blog/external-service/get_facebook_token.html

kenzauros commented:
Okay. Please tell Mr. Masuda to issue a Page Access Token (indefinite one?) following to the blog post you mentioned.
*He is on a business trip.

kenzauros commented:
Except for this, does the workflow work nicely?

jluis9796 commented:
Okay. I will ask Mr. Masuda the token.

The workflow is running but I need to do fixes. For example, the hash comparison returns like the xml is different even though the xml is the same, and check the `jq` command you suggested.

Apart from that, I have some questions.

- To get the info from rss.xml, I am using an external service (https://api.rss2json.com/v1/api.json?rss_url=${XMLDATA}) to get the xml content as an object. Is it okay to use that external service or should I retrieve the info manually?
- Currently I am retrieving only one post, the post with the latest `pubDate` from the `.xml` file. Only one is enough?

kenzauros commented:
> * To get the info from rss.xml, I am using an external service ([https://api.rss2json.com/v1/api.json?rss_url=${XMLDATA}](https://api.rss2json.com/v1/api.json?rss_url=$%7BXMLDATA%7D)) to get the xml content as an object. Is it okay to use that external service or should I retrieve the info manually?

I didn't know the service. The RSS is a public data so basically there is no problem, I think.

" the hash comparison returns like the xml is different" occurs even if you convert the xml to JSON by the service?

> * Currently I am retrieving only one post, the post with the latest `pubDate` from the `.xml` file. Only one is enough?

Yes. It's enough. (so far)
But it's nice to handle the all differential posts.

jluis9796 commented:
> 
> " the hash comparison returns like the xml is different" occurs even if you convert the xml to JSON by the service?
> 

Actually, I confirmed the hash value itself is okay because the value is the same in different runs. So I think the problem is my setting of the cache and the previous value may be empty.

kenzauros commented:
Do you cache a JSON itself or hash only?

jluis9796 commented:
I want to cache the hash only. Currently the workflow for cache looks like below. I need to understand the setting better.

`9f5366decdc92b49a41db4ddccaf4d98c00bd4862efc359b7d9d48de11b3b63e` is the hash value.

```
Run actions/cache@v4
  with:
    path: .cache
    key: blog-post-9f5366decdc92b49a41db4ddccaf4d98c00bd4862efc359b7d9d48de11b3b63e
    enableCrossOsArchive: false
    fail-on-cache-miss: false
    lookup-only: false
    save-always: false
```

jluis9796 commented:
Maybe this is the cause. At the end of the workflow, I see this message. Maybe `.cache` is not a correct value for `path`.

`Warning: Path Validation Error: Path(s) specified in the action for caching do(es) not exist, hence no cache is being saved.`

jluis9796 committed:
modify cache path

kenzauros commented:
I guess the `path` parameter indicates the file path in the job container.
So you must place a text file (to cache) to the path and run the `cache` step.

I think it's simple to place the json downloded and cache it.

jluis9796 commented:
Thanks! I am understanding it better.

jluis9796 committed:
revert changes to use pageID

jluis9796 committed:
add file to cache

jluis9796 committed:
apply jq command to json

jluis9796 committed:
add if condition to workflow steps

jluis9796 committed:
remove console.log

jluis9796 commented:
I fixed the cache to include the downloaded file and now the cache is saved. So, when the .xml content has not changed, the post steps are skiped.
And I replaced the code extracting the json valur with `jq` command.

kenzauros commented on .github/workflows/release-facebook.yml:
```diff
@@ -0,0 +1,90 @@
+      name: Release
+
+    env:
+      BLOG_URL: https://mseeeen.msen.jp/rss.xml
```
This should be `RSS_URL`

jluis9796 replied:
Thanks. I renamed it accordingly.

kenzauros commented on .github/workflows/release-facebook.yml:
```diff
@@ -0,0 +1,90 @@
+
+    env:
+      BLOG_URL: https://mseeeen.msen.jp/rss.xml
+      RSS_FILENAME: rss.xml
```
The extension is `.xml` but the actual content will be JSON.

jluis9796 replied:
I fixed the extension to `.json`.

jluis9796 committed:
fix variable name, fix file extension

jluis9796 committed:
add page access token

jluis9796 committed:
comment if condition for testing

jluis9796 committed:
remove unncessary workflow step

jluis9796 committed:
organize code block

jluis9796 committed:
change API version

jluis9796 committed:
add workflow step to create page access token

jluis9796 committed:
fix workflow step

jluis9796 commented:
I executed the workflow with the access token and the execution seemed to work correctly. The API response was 200 OK.
However, the facebook page (and photos page) seems unaltered. The posted image is the same image of the last post, so maybe the image already existed. I will test it with another image.

![スクリーンショット 2024-05-17 115630](https://github.com/mseninc/blog/assets/10829398/225e6573-f472-4305-9a75-37401b8f3bcd)

jluis9796 committed:
add test image

jluis9796 committed:
add step to post

jluis9796 committed:
add test photo

jluis9796 committed:
set published to false

jluis9796 commented:
I upload photo with `published=true` and got 200 OK. And I uploaded photo with `published=false` and create a post ("https://graph.facebook.com/msenjp/feed).

I don't see those posts in the Facebook page. Does it require to be logged in Facebook or to be a follower? (I don't have Facebook and cannot see it)

Here are screenshot of the result of uploading a photo as a post, and a screenshot of creating a post with an uploaded photo.

![スクリーンショット 2024-05-17 115630](https://github.com/mseninc/blog/assets/10829398/5a864c32-a866-474f-b49d-772c8b7e1c8d)
![スクリーンショット 2024-05-17 134419](https://github.com/mseninc/blog/assets/10829398/1f9d6178-e4ae-4744-87cf-d3e8a5647b1b)

kenzauros commented:
Could you find a way to make photos public?

jluis9796 commented:
Okay. I was looking for this issue. I found one comment in a post saying

> Posts made with the API only show up if the App is public/live. It does not matter if it is for personal use only.

https://stackoverflow.com/questions/60485995/posting-on-my-facebook-page-using-facebook-graph-api-not-showing-in-public

I don't know if this might means some setting required in Facebook.

kenzauros commented:
Ah it means we should make the App (in Facebook) public?

jluis9796 commented:
That is not an official post from Facebook, but according to the post, yes, making the App public should solve it. Although I guess the App is not public for some reason.

kenzauros commented:
Maybe a kind of application for publication is necessary

- [Facebook開発者アカウントでアプリを設定 | Adjust Help Center](https://help.adjust.com/ja/article/set-up-an-app-in-your-facebook-developer-account)

jluis9796 committed:
fix photo url

jluis9796 committed:
add if condition

jluis9796 commented:
For testing, I commented the hash condition (to check if the hash has changed or not) and pushed commits to run the app.
I added again the hash condition.

I tried to add `workflow_dispatch` setting to run the workflow manually, but I cannot see it in GitHub. Maybe because this workflow is not yet in the default branch. If not necessary in the future, I will remove it.

![スクリーンショット 2024-05-17 153549](https://github.com/mseninc/blog/assets/10829398/97a30e47-0664-408f-a7fd-0baa8dc42e65)

kenzauros commented:
Keep `workflow_dispatch` for the future testing.

jluis9796 committed:
comment if condition

jluis9796 committed:
add caption parameter to photo post API

jluis9796 committed:
add quotes to json value

jluis9796 committed:
Test publish photo directly

jluis9796 committed:
fix text breakline

jluis9796 committed:
organize workflow variables

jluis9796 commented:
I replaced RSS url string with the following GitHub actions vars.
```
      RSS_URL: ${{ vars.RSS_FEED_URL }}
```
But I am getting empty value. I would like to confirm if `RSS_FEED_URL` is already defined. Also for `FACEBOOK_PAGE_ID`.

kenzauros commented:
Does the page ID work with `msenjp`?

kenzauros commented:
I have not added them yet. Done.

![20240522_114734](https://github.com/mseninc/blog/assets/13431810/d420c7ca-2ec2-4034-abc2-2bf102951656)

jluis9796 commented:
Yes, `msenjp` is correct for Page ID.

jluis9796 committed:
change variable

jluis9796 commented:
Thanks! I tested and it works.

kenzauros commented:
Maybe extra line feed needed? (I'm not sure)

old

![20240522_115418](https://github.com/mseninc/blog/assets/13431810/c5228200-18fa-4bc9-8548-ba6004d9d40e)

new

![20240522_115425](https://github.com/mseninc/blog/assets/13431810/ef76dc4c-2aa1-4644-b715-14395a8d342a)

jluis9796 commented:
Yes, I noticed that space but it didn't look like to me an extra line. I will add it to see how it looks.

kenzauros commented on .github/workflows/release-facebook.yml:
```diff
@@ -0,0 +1,105 @@
+
+on:
+  workflow_dispatch:
+    inputs:
+      page_id:
+        description: 'Path parameter for /v19.0/{page-id}/photos endpoint'
+        required: false
```
Could you make parameters acceptable from both inputs and vars/secrets?

Refer to https://github.com/mseninc/blog/blob/release/.github/workflows/release-02-preprocess.yml

jluis9796 replied:
Okay.

jluis9796 replied:
I added the parameters to the `workflow_dispatch` inputs and made the value to accept inputs and vars/secrets.
I used `::add-mask::` to mask the secret values (user access token and page access token).
I cannot yet test to input the values in the actions page.

kenzauros replied:
I didn't know `::add-mask::` feature.

Thanks. Ah since it's not yet merged.
Do you think I should merge this?

jluis9796 replied:
I restored the `if` conditions to check hash, and removed this branch from `on` triggers, so now it could be merged.
It does not conflict with other code.

jluis9796 committed:
add extra space to caption text

jluis9796 commented:
I added the extra space and it looks correct.

jluis9796 committed:
add input parameters

jluis9796 committed:
fix secret

jluis9796 committed:
test token input

jluis9796 committed:
use access token from env variable

jluis9796 committed:
mask page access token

jluis9796 committed:
add final fixes

kenzauros commented on .github/workflows/release-facebook.yml:
```diff
@@ -0,0 +1,123 @@
+              photo_id=$(echo $json | jq -r '.id')
+              echo "photo_id=${photo_id}" >> $GITHUB_OUTPUT
+    
+#       - name: Post feed to Facebook using the uploaded photo
+# #        if: ${{ steps.blog-posts-cache.outputs.cache-hit != 'true' }}
+#         run: |
+#             curl -i -X POST \
+#               -d "message=" \
+#               -d "attached_media[0]={"media_fbid":"${{ steps.photo_upload.outputs.photo_id }}"}" \
+#               -d "access_token=${{ steps.token.outputs.page_access_token }}" \
+#               "https://graph.facebook.com/${{ env.PAGE_ID }}/feed"
```
Are these lines needed?

jluis9796 replied:
Actually, it is not needed. I just kept it in case it can be useful later.

kenzauros replied:
Okay. For future reference, tell me what situation can it be useful for?

jluis9796 replied:
This is not a photo upload, but a post with multiple images (`attached_media[0]`, `attached_media[1]`, and so on) and having one single common message.
I tried to use it when I couldn't manage to publish only photos. It worked but later I didn't need it any more because the `/photo` API was enough.

kenzauros replied:
I see. Could you add some comments to describe the use case?

jluis9796 replied:
I added some comments to describe it.

kenzauros commented on .github/workflows/release-facebook.yml:
```diff
@@ -0,0 +1,123 @@
+  push:
+    branches:
+      - release
+      - 317-add-workflow-for-facebook-post
```
I noticed that the `push` (`release`) trigger won't work with this workflow since the RSS won't have been updated yet when the branch is merged.

Use `schedule` trigger. Set 1 pm (JST) everyday.

- [schedule - ワークフローをトリガーするイベント - GitHub Docs](https://docs.github.com/ja/actions/using-workflows/events-that-trigger-workflows#schedule)

jluis9796 replied:
I see. I will change it to use `schedule` you suggest.

jluis9796 replied:
I added the schedule trigger.

jluis9796 committed:
add workflow trigger schedule

jluis9796 committed:
add comments

kenzauros commented:
Could I remove `WIP` from the title and merge this?

jluis9796 commented:
Yes, please.

kenzauros commented:
I'm running the workflow now manually.

By the way, fix the name of the workflow to `Post to Facebook` since `Release` prefix means "regarding to post-releasing process".

kenzauros commented:
It successfully posted a photo. And nothing happens when I re-run it (it works correctly).

![20240522_154845](https://github.com/mseninc/blog/assets/13431810/312ee326-d100-4778-8149-bda650a0d62b)

It seems working good!

jluis9796 commented:
Great to know the dispatch also works well. It should run tomorrow and nothing should happen.

kenzauros commented:
I disabled the old integration on Zapier.

![20240522_155308](https://github.com/mseninc/blog/assets/13431810/592fb6b4-afd0-4f42-a164-44a0711b352a)

kenzauros commented:
> By the way, fix the name of the workflow to `Post to Facebook` since `Release` prefix means "regarding to post-releasing process".

Could you fix this in another branch?

jluis9796 commented:
I did the fix in another PR.
